src/docker_snapshot/container.py

```diff
@@ -109,44 +132,65 @@ def alloc():
     container.start()
 
 
-def dealloc():
+def dealloc() -> None:
+    global container
+
+    if container is None:
+        return
+
     container.stop(timeout=0)
 
 
-def sh(command):
+def sh(command: str) -> str:
     global container
-    code, output = container.exec_run(["sh", "-c", command])
+
+    # TODO typing: fix
+    if container is None:
+        raise RuntimeError("container global not set")
```

Comment:
In general: the only semantic differences on this branch are these types of edits, which don't alter functionality (because the program would've crashed on invalid attribute access anyways).